core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -406,11 +407,28 @@ public static Expression convert(Expression operand, Type fromType,
     if (toPrimitive != null) {
       if (fromPrimitive != null) {
         // E.g. from "float" to "double"
-        return Expressions.convert_(
+        if (toPrimitive == Primitive.BOOLEAN) {
```

Comment:
nit: third level of nested `if` might be a bit much, no worries if refactoring isn't simple/easy.